### PR TITLE
Remove extra logic from file watcher

### PR DESCRIPTION
**Pull request by pakrym**
Remove FILE_WATCHED_ENTRY.
Make server wait for applications to shutdown before serving it.
Should fix failing inprocess stress test too.

### CODE REVIEW

File: src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp [hunk 1]
```diff
@@ -64,8 +64,9 @@ APPLICATION_INFO::GetOrCreateApplication(
         if (m_pApplication->QueryStatus() == RECYCLED)
         {
             LOG_INFO("Application went offline");
-            // Application that went offline
-            // are supposed to recycle themselves
+
+            // Call to wait for application to complete stopping
+            m_pApplication->Stop(false);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp]: nit: `/*fServerInitiated */

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp [hunk 2]
```diff
@@ -45,55 +46,66 @@ FILE_WATCHER::~FILE_WATCHER()
...
+    if (pszDirectoryToMonitor == NULL ||
+        pszFileNameToMonitor == NULL ||
+        pApplication == NULL)
     {
-        hr = HRESULT_FROM_WIN32(GetLastError());
+        DBG_ASSERT(FALSE);
+        return HRESULT_FROM_WIN32(ERROR_INVALID_PARAMETER);
+    }
 
-        CloseHandle(m_hCompletionPort);
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp]: Because I changed it to HandleWrrapper

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp [hunk 3]
```diff
@@ -334,179 +261,67 @@ HRESULT
...
-
-    if (fFileChanged)
+    if (fFileChanged && !_lStopMonitorCalled)
     {
-        //
-        // so far we only monitoring app_offline
-        //
-        _pCallback();
+        // Reference application before 
+        _pApplication->ReferenceApplication();
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp]: Yes, in the QueueUserWorkItem call back.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp [hunk 4]
```diff
@@ -69,6 +69,16 @@ IN_PROCESS_APPLICATION::Stop(bool fServerInitiated)
     HRESULT hr = S_OK;
     CHandle  hThread;
     DWORD    dwThreadStatus = 0;
+
+    SRWExclusiveLock stopLock(m_stateLock);
+
+    if (m_fStopCalled)
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: We want to lock to make other threads wait while Stop is happening.

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp [hunk 5]
```diff
@@ -269,16 +204,8 @@ HRESULT
 
 --*/
 {
-    HRESULT                     hr = S_OK;
-    FILE_NOTIFY_INFORMATION *   pNotificationInfo;
     BOOL                        fFileChanged = FALSE;
 
-    AcquireSRWLockExclusive(&_srwLock);
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp]: file watcher is not shared anymore so I don't see any point in locking this.

**Comment by pakrym** [comment]
https://github.com/aspnet/IISIntegration/issues/1100

File: src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp (hunk 4 above)
Comment (mikeharder) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: It seems like some applications exit early from `Stop()` if `m_fStopCalled` is set, but not all.  Why is this needed now at all?  Why wasn't it needed before this change?

Comment (pakrym) [reply src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp]: Applications that actually do something in stop (stop managed/kill the process) have to prevent double entrance because we call stop both from OnAppOflline and from ApplicationInfo now.

We call Stop from ApplicationInfo to make sure application stops before we start serving app_offline and files are unlocked.

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h [hunk 6]
```diff
@@ -6,124 +6,56 @@
...
+    void StopMonitor();
 
 private:
-    virtual ~FILE_WATCHER_ENTRY();
+    HandleWrapper<NullHandleTraits>               m_hCompletionPort;
+    HandleWrapper<NullHandleTraits>               m_hChangeNotificationThread;
+    volatile   BOOL      m_fThreadExit;
 
-    DWORD                   _dwSignature;
+    DWORD                   _dwSignature{};
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h]: Remvoe?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs [hunk 7]
```diff
@@ -99,15 +102,14 @@ public IISApplication(DeploymentParameters deploymentParameters, ILogger logger)
                         _logger.LogInformation($"Tried to start site, state: {state.ToString()}");
                     }
                 }
-                catch (COMException comException)
+                catch (COMException comException) when (comException.HResult == ERROR_OBJECT_NOT_FOUND)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs]: we will need to be careful of the rebase between our PRs here.

**Comment by pakrym** [comment]
I'm going to merge this. Jenking does not seem to cooperate this morning and other PRs are affected by changes in this one.
